base/loading.jl

```diff
@@ -511,26 +592,55 @@ module_uuid(m::Module) = ccall(:jl_module_uuid, UInt64, (Any,), m)
 
 isvalid_cache_header(f::IOStream) = 0 != ccall(:jl_read_verify_header, Cint, (Ptr{Void},), f.ios)
 
-function cache_dependencies(f::IO)
-    modules = Tuple{Symbol,UInt64}[]
-    files = Tuple{String,Float64}[]
+function parse_cache_header(f::IO)
+    modules = Dict{Symbol,UInt64}()
     while true
         n = ntoh(read(f, Int32))
         n == 0 && break
-        push!(modules,
-              (Symbol(read(f, n)), # module symbol
-               ntoh(read(f, UInt64)))) # module UUID (timestamp)
+        sym = Symbol(read(f, n)) # module symbol
+        uuid = ntoh(read(f, UInt64)) # module UUID (mostly just a timestamp)
+        modules[sym] = uuid
     end
-    read(f, Int64) # total bytes for file dependencies
+    totbytes = ntoh(read(f, Int64)) # total bytes for file dependencies
+    # read the list of files
+    files = Tuple{String,Float64}[]
     while true
         n = ntoh(read(f, Int32))
         n == 0 && break
+        totbytes -= 4 + n + 8
+        if n < 0 # probably means this wasn't a valid file to be read by Base.parse_cache_header
+            error("EOF while reading cache header")
+        end
         push!(files, (String(read(f, n)), ntoh(read(f, Float64))))
     end
+    @assert totbytes == 4 "header of cache file appears to be corrupt"
```

Comment:
How about the following proposal. Can we write a detailed .ji file integrity checker, with as much diagnostic info as possible? Or a diagnostic mode that could be opt in? It could be done here or elsewhere, but if elsewhere it would need to be kept up to date. For the normal path where we don't expect people to be modifying ji files, leave as is.